bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/RefreshingOAuth2CredentialsInterceptor.java

```diff
@@ -220,53 +230,90 @@ protected void checkedStart(Listener<RespT> responseListener, Metadata headers)
    *
    * @throws IOException
    */
-  public void asyncRefresh() throws IOException {
+  ListenableFuture<HeaderCacheElement> asyncRefresh() {
     if (isAppEngine) {
-      syncRefresh();
-    } else if (canRefresh()) {
-      executor.execute(new Runnable() {
-        @Override
-        public void run() {
-          doRefresh();
-        }
-      });
+      SettableFuture<HeaderCacheElement> f = SettableFuture.create();
+      try {
+        f.set(syncRefresh());
+      } catch (Exception e) {
```

Comment:
I'm sure that would be good. I don't think we want to proxy Errors to user code.